rust/pytests/rootcanal.rs

```diff
@@ -0,0 +1,295 @@
+) -> task::JoinHandle<()> {
+    tokio::spawn(async move {
+        loop {
+            select! {
+                _ = cancellation_token.cancelled() => {
+                    // signal copier tasks to close
+                    rx.close();
+                    // print any buffered lines
+                    while let Some(line) = rx.recv().await {
+                        debug!("{line}")
+                    }
+
+                    break;
+                }
+                opt = rx.recv() => {
+                    match opt {
+                        None => break,
+                        Some(line) => debug!("{line}"),
+                    }
+                }
+            }
+        }
+    })
+}
+
+/// Returns the path to a rootcanal binary, downloading it from GitHub if necessary.
+async fn find_rootcanal() -> anyhow::Result<path::PathBuf> {
+    let version = "1.10.0";
+    // from https://github.com/google/rootcanal/releases
+    let (url, zip_bin_path) = match (env::consts::OS, env::consts::ARCH) {
+        ("linux", "x86_64") => ("https://github.com/google/rootcanal/releases/download/v1.10.0/rootcanal-1.10.0-linux-x86_64.zip", "rootcanal-linux-x86_64/bin/rootcanal"),
+        ("macos", "x86_64")  => ("https://github.com/google/rootcanal/releases/download/v1.10.0/rootcanal-1.10.0-macos-x86_64.zip", "rootcanal-macos-x86_64/bin/rootcanal"),
+        _ => {
+            return Err(anyhow!("No Rootcanal binary available for {} {}, sorry", env::consts::OS, env::consts::ARCH));
+        }
+    };
+    let rootcanal_dir = directories::ProjectDirs::from("com", "google", "bumble")
+        .map(|pd| {
+            pd.data_local_dir()
+                .join("test-tools")
+                .join("rootcanal")
+                .join(version)
+                .join("bin")
+        })
+        .ok_or_else(|| anyhow!("Couldn't resolve rootcanal dir"))?;
+    fs::create_dir_all(&rootcanal_dir)?;
+
+    with_dir_lock(&rootcanal_dir.join(".lockdir"), async {
+        let rootcanal_bin = rootcanal_dir.join("rootcanal");
+        if rootcanal_bin.exists() {
+            return Ok(rootcanal_bin);
+        }
+
+        info!("Downloading rootcanal {version} to {:?}", rootcanal_bin);
+
+        let resp_body = reqwest::get(url).await?.bytes().await?;
+        let mut archive = zip::ZipArchive::new(std::io::Cursor::new(&resp_body))?;
+        let mut zip_entry = archive.by_name(zip_bin_path)?;
+        let mut buf = Vec::new();
+        zip_entry.read_to_end(&mut buf)?;
+        fs::write(&rootcanal_bin, buf)?;
+
+        cfg_if! {
+            if #[cfg(unix)] {
+                use std::os::unix::fs::PermissionsExt;
+                fs::set_permissions(&rootcanal_bin, fs::Permissions::from_mode(0o744))?;
+            }
+        }
+
+        Ok(rootcanal_bin)
+    })
+    .await
+}
+
+/// Execute `closure` with a simple directory lock held.
+///
+/// Assumes that directory creation is atomic, which it is on most OS's.
+async fn with_dir_lock<T>(
```

Comment:
Needed since tests run concurrently, and I don't want to sacrifice that as integration-level ones are rather slow